### ISSUE TITLE
allow limiting number of connections per user per profile

### ISSUE THREAD

**Issue by ghost** [eduvpn/vpn-server-api#82]
Now as long as a user has access to a profile, they connect as many times as they want in parallel.

A feature would be to limit the number of simultaneous connects per user per profile.

**Comment by ghost** [comment]
implementation:

- "client connect" has to check currently connected users to see whether a new connection is allowed...
- what to do if the user is not allowed? the client will just fail to connect, there is no backchannel to inform the client (at the moment)

**Comment by sorincocorada** [comment]
We also need such a feature. I have tried to replace the commonName with the userId in CertificatesModule.php but it breaks other things...

**Comment by ghost** [comment]
Hi! 

> We also need such a feature. I have tried to replace the commonName with the userId in CertificatesModule.php but it breaks other things...

What do you need it for specifically? We are trying to find a reason to implement this, so far there isn't really any except "nice to have". The only reason would be if you try to monetize the VPN service and have people pay per connection or something. There may be a (slight) security benefit to allowing only 1 connection per user to the server... if the user can't connect it means they already have some connection active, or an attacker has. But this is a bit of a stretch :)

Maybe you thought about this some more already?

**Comment by ghost** [comment]
Other considerations:

* Should this be a "global" limit, or per "profile"?
* Could this be a global "maxClientConnectPerUser" => 5 to restrict the number of connections per user to 5 for all users? This in order to prevent users going "crazy" with number of connections, but not something "normal" users will ever run into;
* Does there need to be an "override", i.e. user X can connect 5 times, user Y only 1 time?

**Comment by sorincocorada** [comment]
Hi,
Thank your for the quick feedback. 
The main reason for limiting the number of connections to one would be to increase security and to avoid wasting resources by not allowing users to connect from multiple devices simultaneously.
 I was also thinking about assigning users a static IP address, in order to apply further iptables rules to restrict access to designated resources (per user), that's why I have tried to replace the random generated  commonName with the userId...

From my point of view a per profile defined limit would be more than enough, but a global limit would also suffice.
I did not think about monetizing users because we are a public university.

**Comment by ghost** [comment]
Note, I am not trying to say we should not do this, I am trying to get very good concrete reasons and benefits that are not vague for doing this :)

> to increase security

how exactly? threat model? how does limiting users help? Is there another easier way to accomplish the same envisioned security benefits?

> wasting resources by not allowing users to connect from multiple devices simultaneously 

but maybe there is a use case for this? i.e. connected from tablet, laptop, phone? It wouldn't be necessarily "wasting resources"... How many resources would be wasted? Is this a problem in reality? You can always add more VPN server resources if necessary...

> I was also thinking about assigning users a static IP address

This is (also) not supported as we are trying to create a VPN platform that scales really well and is easy to reason about. Having to deal with a fixed IP per users makes this (much) more complicated, especially when dealing with multiple OpenVPN server processes...

**Comment by ghost** [comment]
Another approach would be to limit the number of active certificates per user. This may be easier to implement and have proper error handling in the clients/web interface without the need to try to connect and then fail...

**Comment by sorincocorada** [comment]
Some risks when allowing simultaneous connection from the same user:
- an user may share it credentials with another user and use the connection at the same time;
- if an attacker logs in with stolen credentials, it may be difficult for the legitimate user to detect it;

**Comment by sorincocorada** [comment]
> Another approach would be to limit the number of active certificates per user. This may be easier to implement and have proper error handling in the clients/web interface without the need to try to connect and then fail...

I have tried this approach in a primitive way, by forcing a certificate commonName based on the userId (and on the unique requirement of the commonName in the database). It seems to work for the first user device but afterwards the user has to delete the the authorized application in order to be able to get a new certificate. At the same time the already existing certificate (on the user device) is still valid

**Comment by ghost** [comment]
> I have tried this approach in a primitive way, by forcing a certificate commonName based on the userId (and on the unique requirement of the commonName in the database). It seems to work for the first user device but afterwards the user has to delete the the authorized application in order to be able to get a new certificate. At the same time the already existing certificate (on the user device) is still valid

Good point. That won't work. 

So far it seems that there is a use case for restricting to *1* connection per user, and not an arbitrary amount, like 5, or 2 or 7, but maybe it requires the same amount of code to implement a max 1 connection per user as a max _n_ connections per user.

i think it makes sense to to then do that *globally*, so not per profile, and not per user, but for all users without exceptions. What do you think? That would make implementing it a lot simpler I think. 

An implementation problem that I didn't think about before: if there are 1000s of users connected, on new client connect it has to be checked that the user is not already connected. This would currently have to query all OpenVPN processes which, even though it doesn't take long, might add up... This is not really something that I want to put in the "on connect" flow. So instead it might have to query the database of connected users with entries that don't have a "disconnected_at" entry yet... Pffff

**Comment by sorincocorada** [comment]
I think it doesn't make sense to implement the limitation per user (the effort seems to be high); per profile would be nice to have maybe in the future. A global limit which applies for all users without exceptions seems to the most effective.

> So instead it might have to query the database of connected users with entries that don't have a "disconnected_at" entry yet

Do you mean to check the connection_log sqlite database inside the connect-script?

**Comment by ghost** [comment]
> I think it doesn't make sense to implement the limitation per user (the effort seems to be high)

Yes. The effort is mostly in properly testing it to make sure it works and keeps working...

> A global limit which applies for all users without exceptions seems to the most effective.

Agreed.

> Do you mean to check the connection_log sqlite database inside the connect-script?

Exactly... it is not 100% reliable, i.e. when an OpenVPN server process crashes and restarts a connection is not "closed" in the `connection_log` table, but it may be the best possible with the current way OpenVPN is integrated...

**Comment by sorincocorada** [comment]
According to your recommendations I have added the following sections of code :

[Storage.php.patch.txt](https://github.com/eduvpn/vpn-server-api/files/5649598/Storage.php.patch.txt)
[ConnectionsModule.php.patch.txt](https://github.com/eduvpn/vpn-server-api/files/5649603/ConnectionsModule.php.patch.txt)

The maximum number of connections per user can be limited per profile using the **maxUserConnections** profile variable. There is no error reporting to the user, but the behaviour is the same as when the user account is suspended.

maxUserConnections - Limits the maximum number of simultaneous connections a user is allowed to establish using this profile, no, N/A, N/A

If you think that the proposal makes sense I can also take care of initiating the pull request.
The next step would be to find a solution in order to make sure that the connections are always closed in the connection_log. Maybe to mark all connections closed before starting the openvpn daemon

**Comment by ghost** [comment]
Wow thanks for this! It turns out a bit simpler than I imagined in the end! 

> If you think that the proposal makes sense I can also take care of initiating the pull request.

I think we need to decide also whether the *per profile* or *per server* is best to have. Have to think about this some more. Am also talking to colleagues if this is something we want to support and have any need for. 

>  The next step would be to find a solution in order to make sure that the connections are always closed in the connection_log. Maybe to mark all connections closed before starting the openvpn daemon

This is hard. I'd like to avoid doing something "on startup" as that would require modifying the OpenVPN startup script for example. Until now we just don't anything the "cleanup" the log in case OpenVPN dies (hard). We also more or less depend on this not happening for the connection log itself. Although we do consider it as a possibility and show when a client was "lost" on the log page...  It is tricky...checking the log *might* be "good enough", but also maybe not...

**Comment by ghost** [comment]
eduVPN / Let's Connect! 3.x has now the initial implementation to allow restricting the number of active OAuth clients per user. In 3.x an OAuth client can only have 1 active VPN configuration at a time.

https://github.com/eduvpn/vpn-user-portal/commit/5a943ffb32afb82246f07884eaf09f81b7c2506e

Next up is limiting the number of configurations per user when dealing with manual portal downloads.

**Comment by ghost** [comment]
> Next up is limiting the number of configurations per user when dealing with manual portal downloads.

https://github.com/eduvpn/vpn-user-portal/commit/de8b7dd809392a914a15feef17c08508342784ff
